### PR TITLE
[WIP] Check replication lag

### PR DESCRIPTION
**Pull request by UnderGreen**

### CODE REVIEW

File: pkg/postgresql/utils.go [hunk 1]
```diff
@@ -263,6 +278,31 @@ func GetPGState(ctx context.Context, replConnParams ConnParams) (*cluster.Postgr
...
+	// Add "replication=1" connection option
+	replConnParams["replication"] = "1"
+	db, err := sql.Open("postgres", replConnParams.ConnString())
+	if err != nil {
+		return 0, err
+	}
+	defer db.Close()
+
+	rows, err := Query(ctx, db, "select pg_replication_lag()")
+	if err != nil {
```
Comment (UnderGreen) [inline pkg/postgresql/utils.go:289]: @sofuture I don't understand why here I have error:
```
2017-01-20 11:49:14.451 UTC [158]: [1-1] user=repluser,db=[unknown] ERROR:  syntax error
2017-01-20 11:49:14.452023 [keeper.go:420] E | keeper: error getting replication lag: pq: syntax error
```
I thought it was issue with case construction, I changed this line to call only postgres function and again have the error.

Comment (UnderGreen) [reply pkg/postgresql/utils.go:289]: Changed replConnParams to localParams and error changed :)
`2017-01-20 15:13:04.757046 [keeper.go:420] E | keeper: error getting replication lag: sql: Scan error on column index 0: converting driver.Value type <nil> ("<nil>") t
`

Comment (UnderGreen) [reply pkg/postgresql/utils.go:289]: ahh, need to work with NULL return values from query

**Comment by sofuture** [comment]
Lookin good to me!

File: cmd/sentinel/sentinel.go [hunk 2]
```diff
@@ -242,6 +242,9 @@ func (s *Sentinel) GetBestStandby(cv *cluster.ClusterView, keepersState cluster.
 		if k.PGState.ReplicationLag >= s.clusterConfig.MaxReplicationLag {
 			log.Debugf("ignoring node since its replication lag (%d) more than maximum possible lag (%d)", keepersState[id].PGState.ReplicationLag, s.clusterConfig.MaxReplicationLag)
 		}
+		if (masterState.PGState.XLogPos - k.PGState.XLogPos) >= uint64(s.clusterConfig.MaxReplicationLagB) {
+			log.Debugf("ignoring node since its replication lag in bytes (%d) more than maximum possible lag (%d)", (masterState.PGState.XLogPos - k.PGState.XLogPos), s.clusterConfig.MaxReplicationLagB)
```
Comment (klizhentas) [inline cmd/sentinel/sentinel.go]: I would make this a warning

File: cmd/keeper/keeper.go [hunk 3]
```diff
@@ -451,6 +452,14 @@ func (p *PostgresKeeper) GetPGState(pctx context.Context) (*cluster.PostgresStat
 		}
 		pgState.ReplicationLag = replicationLag
 
+		ctx, cancel = context.WithTimeout(pctx, p.clusterConfig.RequestTimeout)
+		role, err := pg.GetRole(ctx, p.getLocalConnParams())
+		defer cancel()
+		if err != nil {
+			return nil, fmt.Errorf("error getting replication lag: %v", err)
```
Comment (klizhentas) [inline cmd/keeper/keeper.go]: can you start using package `trace` instead of built-in `fmt` as we need to track origins of error.

Comment (UnderGreen) [reply cmd/keeper/keeper.go]: Yeah, but only for new or changed functions. I don't have a lot of time to change it everywhere.

File: pkg/postgresql/utils.go (hunk 1 above)
Comment (klizhentas) [reply pkg/postgresql/utils.go:289]: https://marcesher.com/2014/10/13/go-working-effectively-with-database-nulls/

Comment (UnderGreen) [reply pkg/postgresql/utils.go:289]: Sure, I know about it.
https://github.com/gravitational/stolon/pull/26/files#diff-d8eea19f0a5ff785d034ea580f183dccR294

**Comment by UnderGreen** [comment]
@sofuture 
```
2017-01-23 06:59:05.424310 [sentinel.go:527] I | sentinel: master is failed
2017-01-23 06:59:05.424342 [sentinel.go:538] I | sentinel: trying to find a standby to replace failed master
2017-01-23 06:59:05.424379 [sentinel.go:221] D | sentinel: id: 1b72501e, k: (*cluster.KeeperState)Keeper with ID 1b72501e is healthy and listens on 172.28.128.14:5431
2017-01-23 06:59:05.424397 [sentinel.go:246] W | sentinel: ignoring node since its replication lag in bytes (2749321816) more than maximum possible lag (262144)
2017-01-23 06:59:05.424422 [sentinel.go:221] D | sentinel: id: 461281d5, k: (*cluster.KeeperState)Keeper with ID 461281d5 is healthy and listens on 172.28.128.16:5431
2017-01-23 06:59:05.424438 [sentinel.go:243] W | sentinel: ignoring node since its replication lag (1669) more than maximum possible lag (600)
2017-01-23 06:59:05.424518 [sentinel.go:221] D | sentinel: id: fb7adbac, k: (*cluster.KeeperState)Keeper with ID fb7adbac is not healthy and listens on 172.28.128.13:5
2017-01-23 06:59:05.424542 [sentinel.go:223] W | sentinel: ignoring node "fb7adbac" since it's the current master
2017-01-23 06:59:05.424571 [sentinel.go:221] D | sentinel: id: 10c4dd55, k: (*cluster.KeeperState)Keeper with ID 10c4dd55 is healthy and listens on 172.28.128.15:5431
2017-01-23 06:59:05.424586 [sentinel.go:243] W | sentinel: ignoring node since its replication lag (1671) more than maximum possible lag (600)
2017-01-23 06:59:05.424603 [sentinel.go:544] I | sentinel: electing new master: "461281d5"
```
It works as expected.
1b72501e I stopped before writing more than 3 gb data.
10c4dd55 - this node I added after wrote the data and stopped in 2-3 minutes.

**Comment by sofuture** [comment]
:shipit:

**Pull request merged**